sklearn/impute.py

```diff
@@ -321,3 +379,304 @@ def transform(self, X):
+        """Impute all missing values in X.
+
+        Parameters
+        ----------
+        X : {array-like}, shape = [n_samples, n_features]
+            The input data to complete.
+
+        Returns
+        -------
+        X : {array-like}, shape = [n_samples, n_features]
+            The imputed dataset.
+        """
+
+        # Import here to avoid circular import
+        from .neighbors.base import _get_weights
+
+        check_is_fitted(self, 'statistics_')
+        force_all_finite = False if self.missing_values in ["NaN",
+                                                            np.nan] else True
+        X = check_array(X, accept_sparse=False, dtype=FLOAT_DTYPES,
+                        force_all_finite=force_all_finite, copy=self.copy)
+        # Check for +/- inf
+        if np.any(np.isinf(X)):
+            raise ValueError("+/- inf values are not allowed in data to be "
+                             "transformed.")
+
+        # Get fitted data and ensure correct dimension
+        fitted_X = self._fitted_neighbors._fit_X
+        n_rows_fit_X, n_cols_fit_X = fitted_X.shape
+        n_rows_X, n_cols_X = X.shape
+
+        if n_cols_X != n_cols_fit_X:
+            raise ValueError("Incompatible dimension between the fitted "
+                             "dataset and the one to be transformed.")
+        mask = _get_mask(X, self.missing_values)
+
+        row_total_missing = mask.sum(axis=1)
+        if not np.any(row_total_missing):
+            return X
+
+        # Check for excessive missingness in rows
+        bad_rows = row_total_missing > (mask.shape[1] * self.row_max_missing)
+        if np.any(bad_rows):
+            warnings.warn(
+                "There are rows with more than {0}% missing values. The "
+                "missing features in these rows are imputed with column means."
+                    .format(self.row_max_missing * 100))
+            X_bad = X[bad_rows, :]
+            X = X[~bad_rows, :]
+            mask = mask[~bad_rows]
+            row_total_missing = mask.sum(axis=1)
+        row_has_missing = row_total_missing.astype(np.bool)
+
+        if np.any(row_has_missing):
+
+            # Mask for fitted_X
+            mask_fx = _get_mask(fitted_X, self.missing_values)
+
+            # Get row index of missing and distance from donors
+            n_adj_samples, _ = X[row_has_missing].shape
+            dist = pairwise_distances(X,
+                                      fitted_X,
+                                      metric=self.metric,
+                                      squared=False,
+                                      missing_values=self.missing_values)
+
+            # For each column, find and impute missing
+            for c in range(n_cols_X):
+                if not np.any(mask[:, c], axis=0):
+                    continue
+                # Row index for receivers and potential donors (pdonors)
+                receivers_row_idx = np.where(mask[:, c])[0]
```

Comment:
This is very nested. Can we pull it out into a helper to clarify its interface and perhaps to make it easier to test?